example/admin/data_init.go

```diff
@@ -17,6 +18,48 @@ func EmptyDB(db *gorm.DB, tables []string) {
 	}
 }
 
+// ErasePublicUsersData erase all non-admin users but preserve the following three users
+// qor@the-plant.com
+// demo-editor@the-plant.com
+// demo-viewer@the-plant.com
+func ErasePublicUsersData(db *gorm.DB) {
+	reservedAccount := []string{
+		"qor@the-plant.com",
+		"demo-editor@the-plant.com",
+		"demo-viewer@the-plant.com",
+	}
+
+	var err error
+	var adminRoleID int
+	// obtain the admin role ID
+	if err = db.Table("roles").Where("name = ?", models.RoleAdmin).
+		Pluck("id", &adminRoleID).Error; err != nil {
+		panic(fmt.Errorf("failed to obtain the admin role ID! %v", err))
+	}
+
+	// subQuery for finding the ids of these demo users
+	subQuery := db.Table("users").Select("id").Where("account in (?)", reservedAccount)
+	// obtain the user ids to be reserved
+	var reservedUserIds []int
+	err = db.Table("user_role_join").Group("user_id").
+		Having("user_id IN (?) or COUNT(CASE WHEN role_id = (?) then 1 end)=1", subQuery, adminRoleID).
+		Pluck("user_id", &reservedUserIds).Error
+	if err != nil {
+		panic(fmt.Sprintf("failed to obtain the user ids to be retained! %v", err))
+	}
+
+	// First delete the data in the user_role_join table, then delete the data in the users table.
+	// Due to foreign key constraints, it is not possible to delete data from the users table first.
```

Comment:
It's not so important here, but you can avoid problems with constraints by using transactions. The transactions are not enforced until the transaction is committed.

An alternative is to use `ON DELETE CASCADE` on the key, so that the database automatically deletes user role associations when the related users are deleted. If you do this, then this whole function can reduce to `DELETE FROM users where email not in (?)`

Reply:
Thank you for pointing out the potential issue. Since I am not sure if ON DELETE CASCADE was specified when creating the user_role_join table in the database, I will create a new PR that uses transactions to ensure database consistency.